### ISSUE TITLE
2023.12.03 version broke XTB - see link in description

### ISSUE THREAD

**Issue by mtnmuncher** [Innofactor/PluginRegistration#132]
https://github.com/MscrmTools/XrmToolBox/issues/1308

**Comment by ingerAH-PP** [comment]
I experience the same issue when installing the latest version of the tool:

![image](https://github.com/Innofactor/PluginRegistration/assets/51701932/8f5a7e50-b1ec-4ee7-a093-34a328dded78)

XrmToolBox version: v1.2023.10.67


Error details:

```
Error Time: 2023-12-11 13:51:20.578
System.Reflection.ReflectionTypeLoadException
Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
mscorlib
at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at System.ComponentModel.Composition.Hosting.AssemblyCatalog.get_InnerCatalog()
   at System.ComponentModel.Composition.Hosting.AssemblyCatalog.GetExports(ImportDefinition definition)
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.ComponentModel.Composition.Hosting.CatalogExportProvider.InternalGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExports(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.CatalogExportProvider.GetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.AggregateExportProvider.GetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.CompositionContainer.GetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TryGetExports(ExportProvider provider, ComposablePart part, ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImportSubset(PartManager partManager, IEnumerable`1 imports, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImportsStateMachine(PartManager partManager, ComposablePart part)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImports(PartManager partManager, ComposablePart part, Boolean shouldTrackImports)
   at System.ComponentModel.Composition.Hosting.ImportEngine.SatisfyImports(ComposablePart part)
   at System.ComponentModel.Composition.Hosting.CompositionServices.TryInvoke(Action action)
   at System.ComponentModel.Composition.Hosting.ComposablePartExportProvider.Compose(CompositionBatch batch)
   at XrmToolBox.PluginManagerExtended.LoadParts(Boolean isRetry)
   at XrmToolBox.PluginManagerExtended.LoadParts(Boolean isRetry)
   at XrmToolBox.PluginManagerExtended.RescanIfRequired()
   at XrmToolBox.PluginManagerExtended.Initialize(Form parentForm)
   at XrmToolBox.New.PluginsForm2..ctor()
   at XrmToolBox.New.NewForm..ctor(String[] args)
   at XrmToolBox.Program.Main(String[] args)
```

**Comment by devlinl** [comment]
Same here.

**Comment by davidmartiindale** [comment]
Same issue for me

**Comment by imranakram** [comment]
@mtnmuncher @davidmartiindale  Have you tried updating XTB to the latest version?

@MscrmTools  Seems to be working fine with the latest version, any ideas what could cause this issue?

**Comment by ingerAH-PP** [comment]
I tried with the latest version of XrmToolBox: 1.2023.12.68
Plugin Registration version: 3.2023.12.3

![image](https://github.com/Innofactor/PluginRegistration/assets/51701932/79262d9a-2a56-4b14-b9d7-ff263d9af172)

```
Error Time: 2023-12-12 11:10:09.399
System.Reflection.ReflectionTypeLoadException
Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
mscorlib
at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at System.ComponentModel.Composition.Hosting.AssemblyCatalog.get_InnerCatalog()
   at System.ComponentModel.Composition.Hosting.AssemblyCatalog.GetExports(ImportDefinition definition)
   at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
   at System.ComponentModel.Composition.Hosting.CatalogExportProvider.InternalGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExports(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.CatalogExportProvider.GetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.AggregateExportProvider.GetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.CompositionContainer.GetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ExportProvider.TryGetExportsCore(ImportDefinition definition, AtomicComposition atomicComposition, IEnumerable`1& exports)
   at System.ComponentModel.Composition.Hosting.ExportProvider.GetExports(ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TryGetExports(ExportProvider provider, ComposablePart part, ImportDefinition definition, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImportSubset(PartManager partManager, IEnumerable`1 imports, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImportsStateMachine(PartManager partManager, ComposablePart part)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImports(PartManager partManager, ComposablePart part, Boolean shouldTrackImports)
   at System.ComponentModel.Composition.Hosting.ImportEngine.SatisfyImports(ComposablePart part)
   at System.ComponentModel.Composition.Hosting.CompositionServices.TryInvoke(Action action)
   at System.ComponentModel.Composition.Hosting.ComposablePartExportProvider.Compose(CompositionBatch batch)
   at XrmToolBox.PluginManagerExtended.LoadParts(Boolean isRetry)
   at XrmToolBox.PluginManagerExtended.LoadParts(Boolean isRetry)
   at XrmToolBox.PluginManagerExtended.RescanIfRequired()
   at XrmToolBox.PluginManagerExtended.Initialize(Form parentForm)
   at XrmToolBox.New.PluginsForm2..ctor()
   at XrmToolBox.New.NewForm..ctor(String[] args)
   at XrmToolBox.Program.Main(String[] args)
```

**Comment by davidmartiindale** [comment]
I resolved this issue by updating other plugin tools I use before installing, specifically Early Bound Generator & Plugin Trace Viewer

**Comment by mtnmuncher** [comment]
I am already using the latest version of XrmToolBox: 1.2023.12.68

Since I am already using the latest Plug-in trace viewer and do not use Early Bound Generator, I don't understand how updating other plug-ins will resolve the issue. 

If there are, in fact, prerequisite plug-ins, please provide the exact list. If not, please provide a link to download the 'previous' version of Plug-in Registration Tool.

**Comment by mtnmuncher** [comment]
As per previous instructions, I downloaded 3.3023.11, and was able to successfully use this version from inside XTB. 

Out of curiosity, I tried the following update path:

1. Successfully updated to the 3.2023.12.2. 
2. After installing 3.2023.12.2, I was able to successfully update to 3.2023.12.3

Please try this approach and let me know if this works for others.

**Comment by mtnmuncher** [comment]
Steps I performed

1. installed Early Bound Generator (as well as V2) plug-ins. Even though I don't use these, someone mentioned they installed these to get stuff working.
2. installed version 3.3023.11 of PRT
3. I was able to update directly to version 3.2023.12.3 of PRT.

**Comment by devlinl** [comment]
> Steps I performed
> 
> 1. installed Early Bound Generator (as well as V2) plug-ins. Even though I don't use these, someone mentioned they installed these to get stuff working.

Ah, I missed that bit from earlier... this got everything working for me.

**Comment by daryllabar** [comment]
@imranakram, the error is that the Microsoft.Bcl.AsyncInterfaces V6 assembly couldn't be found.  The reason that installing the EBGV2 resolves this, is that the latest version of the XTB will search in the plugin sub-directories for any other dll that it needs, and it finds the EBGV2 version of the dll.  

The solution is to either ILMerge this dll (yuck) or move away from that entirely to putting all dependent assemblies in the XTB\PLugins\PluginRegistration sub folder.

**Comment by YoYoFreakCJ** [comment]
As a temporary workaround you can download the Microsoft.Bcl.AsyncInterfaces.dll (for example from the [official NuGet package](https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces/6.0.0), make sure to actually pick the version 6.0.0 and **not** the most recent version) and simply put it in the %AppData%\MscrmTools\XrmToolBox\Plugins folder along with all the plugins. This will allow the Plugin Registration Tool to run again.

**Comment by imranakram** [comment]
This should be resolved now with version `3.2023.12.12`

**Comment by imranakram** [comment]
Closing this now, feel free to reopen in case you encounter any relevant issues.
